Implement backward passes for llama with small training llama from scratch example

xaedes opened the pull request:
Training a llama directly with ggml would be really nice.

For this I implemented the backward passes required for the llama model, tested them with test_grad0 from ggml repo and trained a small llama from scratch to output a sinus wave.

Also see the more detailed discussion in https://github.com/ggerganov/ggml/issues/8#issuecomment-1518465663

List of all new operations that I had to add:
- GGML_OP_ADD1 : Could be replaced with add(X,repeat(Y,X)) but it is faster when repeat can be avoided.
- GGML_OP_ACC : Necessary for view backward pass. This adds src1 to view(src0, nb, offset) and returns tensor of shape src0. 
- GGML_OP_SET : **(this is new)** Necessary for propagating gradients through kv cache. Instead of copying to kv cache this function sets the values in the kv cache viewed with offsets and strides and returns a tensor representing the modified kv cache. This can also be inplace returning a view to the modified kv cache.
- GGML_OP_LOG : **(this is new)** Necessary for cross entropy loss
- GGML_OP_SUM_ROWS : Necessary for repeat backward pass: Reduces rows by summing them. shape[a,b,c,d] -> shape[1,b,c,d]
- GGML_OP_SILU_BACK : Necessary for silu backward pass
- GGML_OP_RMS_NORM_BACK : Could also be implemented using primitives, at the cost of performance.
- GGML_OP_GET_ROWS_BACK : Necessary for get_rows backward pass: Adds src0[i] rows to opt0[src1[i]] rows, returning a tensor of shape opt0. 
- GGML_OP_DIAG : Necessary for softmax backward pass, alternative would have been to implement SOFTMAX_BACK directly, but DIAG is at least usable for other stuff. It turns rows into diagonal matrices.
- GGML_OP_DIAG_MASK_ZERO : Necessary for diag_mask_inf backward pass
- GGML_OP_ROPE_BACK : Necessary for rope backward pass.

Notable other changes:
- add inplace and non-inplace variants for scale, diag_mask_inf, soft_max and rope
- fix sub, mul and div functions to work correctly with transposed tensor, uses the same logic as in add:
- fix ggml_forward_add functions to work correctly with transposed tensors. uses the same logic as in ggml_compute_forward_add_q_f32, but make it consistent across all ggml_compute_forward_add_... functions. this also slightly changes the mem access pattern of the different threads to work as in ggml_compute_forward_add_q_f32. 
- de-duplicate ggml_forward_dup code taking care of contiguous tensors of same type. with this we can duplicate tensor of any type as long as they are contiguous. the function is used in dup, get_rows_back and diag_mask (when not inplace).
- there are some *maybe* too verbose comments including step-by-step derivation of gradients that could (or should?) be cleaned away. 
- **(this is new)** I added 1d and 4d functions for ggml_reshape and ggml_view.

The performance of various parts of the training could be improved, especially a fast ggml_out_prod could help speeding up the matrix multiplication backward pass.

There are two additional test files, one for testing gradients taken from ggml repo and one small test for testing optimization in general.

Exemplary training of a small llama model is demonstrated in the self-contained baby-llama example, where it is trained to output a sinus wave.

![](https://i.imgur.com/YGD24GX.gif)

Some notes on first training tests:
- lbfgs optimizer is faster and trains better than adam
- target logits should represent NEXT token, not the current
- target logits -1 & +1 work much better than 0 & +1
- I think adding a BOS token also helped improving the training
- trained with cross entropy loss gave worse generation results than trained with summed squared logit difference loss

A parallel batched forward function would probably be a good improvement. Training on multiple examples in (parallel) batch really seems to improve the training, but currently I can only do that by calling the forward function multiple times with different input data, which costs a lot of nodes in the computation graph, especially since backward pass is necessary as well. 
The batches could be stored in another dimension of the tensors. It would probably just require some reshapes and view operations to make it work.

I did not look into training a LoRa finetune yet, but all the necessary machinery for that seems to be working.

xaedes committed:
norm & rms_norm can not be threaded:

after investigation rms norm for quite some time I come to the conclusion that neither norm, nor rms_norm can be threaded, because we need mean over all items, not just of the slices each thread sees.

xaedes committed:
remove already resolved TODO

xaedes committed:
implement backward pass of ggml_rope and ggml_rope_back

xaedes committed:
use GGML_PRINT_DEBUG for debug messages which will otherwise flood the console

xaedes committed:
test both gradients of mul_mat

xaedes committed:
implement backward pass for ggml_get_rows and for new operation ggml_get_rows_back

xaedes committed:
add test-grad0.c

xaedes committed:
disable graph dot export as it floods console

xaedes committed:
bug fixes for silu_back

xaedes committed:
successfully test silu backward

xaedes committed:
bug fix for scale backward pass

use sum instead of mean for gradient of scalar scale parameter

xaedes committed:
successfully test scale backward

xaedes committed:
improve performance of sum backward pass

use add1(x,y) instead of add(x,repeat(y,x))

xaedes committed:
improve performance of sqr backward pass

use scale(x,y) instead of mul(x,repeat(y,x))

xaedes committed:
successfully test rope backward

xaedes committed:
bug fix for cpy backward pass

xaedes committed:
successfully test cpy backward

xaedes committed:
bug fix for reshape backward pass

xaedes committed:
successfully test reshape backward

xaedes committed:
add test-opt.c

this uses ggml_opt to train a,b for minimal e=sum(sqr(c - a*b)) for random initial a,b,c

xaedes committed:
correctly implement softmax backward pass using new operation ggml_diag

ggml_diag constructs diagonal matrices with entries.
ggml_diag(shape[a,1,c,d]) -> shape[a,a,c,d]

xaedes committed:
successfully test soft_max backward

xaedes committed:
align shape annotations

xaedes committed:
add shape annotations for llama

xaedes committed:
de-duplicate ggml_forward_dup code taking care of contiguous tensors of same type.

with this we can duplicate tensor of any typ as long as they are contiguous.

xaedes committed:
bug fix for add_at forward

required for view backward pass

src0 values must be copied to dst, because during addition we don't touch all dst elements in contrast to the normal add function.

xaedes committed:
minor code format improvement

xaedes committed:
fix ggml_forward_add functions to work correctly with transposed tensors

uses the same logic as in ggml_compute_forward_add_q_f32, but make it consistent across all ggml_compute_forward_add_... functions.
this also slightly changes the mem access pattern of the different threads to works as in ggml_compute_forward_add_q_f32.

xaedes committed:
fix ggml_forward_add1 functions to work correctly with transposed tensors

uses the same logic as in ggml_compute_forward_add1_q_f32, but make it consistent across all ggml_compute_forward_add1_... functions.
this also slightly changes the mem access pattern of the different threads to works as in ggml_compute_forward_add1_q_f32.

xaedes committed:
test-grad0.c : add print_elements to help with debugging

xaedes committed:
successfully test permute backward

xaedes committed:
some minor test-grad0 fixes

xaedes committed:
fix sub, mul and div functions to work correctly with transposed tensors

uses the same logic as in add

xaedes committed:
implement ggml_cont backward pass

xaedes committed:
successfully test transpose backward and permute for all permutations

also test sub, mul and div up to max n_dims

xaedes committed:
test-grad0.c add TODO for view_2d and view_3d

add_at (required for view backward pass) is a bit tricky for n_dims > 1.

xaedes committed:
fix comments

xaedes committed:
successfully test view backward

xaedes committed:
successfully test diag_mask_inf and diag_mask_zero backward

xaedes committed:
test-grad0 : fix test for div

nargs and ndims was swapped, corrupting the stack

xaedes committed:
fix diag_mask to work with non-inplace input

xaedes committed:
move dup call into the actual add_at functions

xaedes committed:
fix get rows backward pass

xaedes committed:
successfully test get_rows backward

xaedes committed:
fix view backward pass

add nb parameters to add_at like in view.
together with offset they define how to view dst and src0 during the add_at operation.

xaedes committed:
successfully test backward pass of view_1d, view_2d and view_3d

xaedes committed:
fix backward pass for rms_norm

I would have used formulas from other frameworks, but they differed so I could not decide which is correct.
Instead it was derived here in comment using manual forward-backward automatic differention of rms_norm and simplification.

xaedes committed:
successfully test backward pass of rms_norm

some tests may fail when gradients are large.
could not find a satisfying configuration to check for abs error and relative error that passes all tests while still actually testing the results with tight enough error bounds.
when looking at the values the "failed" tests look actually ok. for example:

rms_norm: ndims=2, i=0, k=2, x0=0.000153, xm=0.000053, xp=0.000253, f0=0.278594, f1=0.086213, g0=961.905457, g1=966.064941, eps=0.000100, error_abs=4.159485, error_rel=0.004324

it is due to the test logic in check_gradients that they fail.

xaedes committed:
add todos for llama backward pass

- implementation for ADD1 backward pass should probably use sum instead of mean (but this backward pass is not required)
- repeat is not yet tested and looks like it only works for single element src0 inputs.

xaedes committed:
add operation ggml_sum_rows

ggml_sum_rows(shape[a,b,c,d]) -> shape[1,b,c,d]

xaedes committed:
fix backward pass for repeat

requires ggml_sum_rows

xaedes committed:
successfully test backward pass of repeat

xaedes committed:
implement 5 of 6 missing backward pass operations used by llama

- GGML_OP_DIAG_MASK_INF
- GGML_OP_GET_ROWS
- GGML_OP_RMS_NORM
- GGML_OP_SILU
- GGML_OP_SOFT_MAX

add necessary ggml operations GGML_OP_ADD1, GGML_OP_SILU_BACK, GGML_OP_RMS_NORM_BACK, GGML_OP_DIAG_MASK_ZERO, and GGML_OP_ROPE_BACK

GGML_OP_ADD1 is necessary to add a scalar value in the backward pass of GGML_OP_SOFT_MAX
GGML_OP_ADD1 could also be replaced by using GGML_OP_ADD and GGML_OP_REPEAT, but the performance would be worse. additionally GGML_OP_REPEAT will return unexpected value when the the input to GGML_OP_SOFT_MAX contains only a single scalar. in this case GGML_OP_REPEAT will not return the value that should be repeated (src1) but the value which shape the result should take (src0). So in this case it can not replace GGML_OP_ADD1.

GGML_OP_SILU_BACK, GGML_OP_RMS_NORM_BACK and GGML_OP_ROPE_BACK are necessary for backward pass of GGML_OP_SILU, GGML_OP_RMS_NORM and GGML_OP_ROPE. The backward pass for these functions cannot be easily composed of existing operations. Since the backward pass builds a computation graph we need operations forward pass implementations of the the required backward passes. Sounds a bit confusing at first, I know...

GGML_OP_DIAG_MASK_ZERO is necessary for backward pass of GGML_OP_DIAG_MASK_INF.

Some operations where previously inplace-only. for backward pass there needs to be non-inplace variants.
staying consistent with other operations that have non-inplace and inplace variants, the operations are changed to non-inplace and
functions with "_inplace" are added which are inplace.
in llama we need to call the inplace variants so that it is implemented as before.
for llama backward pass we need to use the non-inplace variants.

still not completely implemented backward passes for llama:

- GGML_OP_ROPE: needs forward pass for GGML_OP_ROPE_BACK
- GGML_OP_GET_ROWS: only necessary for tokenizer

xaedes committed:
implement 8 of 14 missing backward pass operations used by llama

- GGML_OP_ADD_AT
- GGML_OP_CPY
- GGML_OP_MUL_MAT (src0.grad)
- GGML_OP_PERMUTE
- GGML_OP_RESHAPE
- GGML_OP_SCALE
- GGML_OP_TRANSPOSE
- GGML_OP_VIEW

implement additional ggml operation GGML_OP_ADD_AT, which is necessary for backward pass of GGML_OP_VIEW.

this operation adds src1 to src0 with data offset, i.e. to view(src0, ..., offset).
the values are return in a tensor size of src0. values outside of [data+offset:data+offset+nbytes(src1)] are just the original values from src0.

still missing backward passes for llama:

- GGML_OP_DIAG_MASK_INF
- GGML_OP_GET_ROWS
- GGML_OP_RMS_NORM
- GGML_OP_ROPE
- GGML_OP_SILU
- GGML_OP_SOFT_MAX

xaedes committed:
add missing GGML_OP_SUM_ROWS

xaedes committed:
update quantization types in switch-case of add_at and add1

xaedes committed:
fix ggml_compute_forward_dup_same_cont for when nelements < nthreads

when more threads are used than elements exist ie1 was less than ie0, resulting in invalid negative byte count argument in memcpy

xaedes committed:
add baby-llama example training a very small llama model from scratch to output a sinusoidal wave.

had to increase maximum number of optimization parameters to train from scratch.

xaedes committed:
fix softmax in baby-llama example

xaedes committed:
switching from training with adam to lbfgs produces much better results in the baby-llama example

xaedes committed:
train with two examples, creating new tensors each time..

xaedes committed:
fix bug when using ggml_opt to optimize params in one context and use a renewable context for eval and opt

when not keeping gradients of model parameters they are overwritten by tensors created by opt, which may be invalid after opt context is renewed.
so we need to keep the original gradients and make dups for opt

xaedes committed:
train on multiple examples, generate & print tokens with trained model afterwards

ctx0 for evaluation and optimization is renewed for each sample

xaedes committed:
add ggml_reshape_1d, ggml_reshape_4d and ggml_view_4d

xaedes committed:
fix soft_max backward pass for input->ne[1] != 1

xaedes committed:
add ggml_log operation necessary for cross entropy loss

xaedes committed:
add test for ggml_log gradients

xaedes committed:
implement backward pass for ggml_sum_rows, necessary for cross entropy loss

xaedes committed:
implement ggml_repeat support for rank > 2 tensors

xaedes committed:
add test for ggml_sum_rows gradients

xaedes committed:
fix training get_example_targets

predict the next token, not the current token!

xaedes committed:
add square_error_loss and cross_entropy_loss functions

xaedes committed:
optimize loss over multiple samples

this increases computation graph, need parallel batched forward for more efficiency.

xaedes committed:
fix backward pass for add_at and change arguments to have same order as in view

xaedes committed:
add ggml_set(ctx, a, b) to set b in view of a and return modified a

necessary to set values into kv_self cache and properly propagate the gradients

xaedes committed:
fix kv_self gradients for training

use ggml_set instead of ggml_cpy to set kv_self cache with properly propagating gradients

xaedes committed:
replace inplace operations for training with copying operations to allow gradient propagation

xaedes committed:
add GGML_ASSERT to catch ggml_rope and back value errors

xaedes committed:
add trainable lora-only model with all big matrices C split into A,B with A*B=C

this is not a lora-finetune, but the whole model changed to have only low-rank "lora" matrices.

training this instead of the normal model resulted in much worse results though...

xaedes committed:
vastly improve training results

instead of logit targets 0 and 1 use -1 and +1.

xaedes committed:
shorten code using a variable

xaedes committed:
change name of GGML_OP_ADD_AT to GGML_OP_ACC

xaedes committed:
Merge branch 'master' into train-example

# Conflicts:
#	ggml.c
#	llama.cpp

xaedes committed:
smaller default values for baby llama model parameters

xaedes committed:
update static assert of GGML_OP_COUNT

xaedes committed:
remove shape annotations in llama_eval_internal

xaedes committed:
revert disabling of threading for rms_norm and norm

xaedes committed:
rename print functions in baby-llama example

xaedes committed:
fix call to ggml_set_name

xaedes committed:
add missing include for strcmp, etc

xaedes committed:
remove trailing whitespace

xaedes committed:
reduce number of test-grad0 iterations

avoid exceeding timeout of automated tests

xaedes committed:
remove busy loop that was used as sleep for slower sinus wave generation

xaedes committed:
disable slow tests grad0 and opt to avoid exceeding timeouts

prusnak commented on examples/baby-llama/baby-llama.cpp:
```diff
@@ -0,0 +1,1354 @@
+#include "ggml.h"
+#include <vector>
+#include <assert.h>
```
C++
```suggestion
#include <cassert>
```

prusnak commented on examples/baby-llama/baby-llama.cpp:
```diff
@@ -0,0 +1,1354 @@
+#include <vector>
+#include <assert.h>
+#include <random>
+#include <string.h>
```
C++
```suggestion
#include <cstring>
```

prusnak commented on examples/baby-llama/baby-llama.cpp:
```diff
@@ -0,0 +1,1354 @@
+#include <random>
+#include <string.h>
+
+#undef MIN
+#undef MAX
+#define MIN(a, b) ((a) < (b) ? (a) : (b))
+#define MAX(a, b) ((a) > (b) ? (a) : (b))
```
Isn't it better to just use `std::min` and `std::max` in C++ rather than this?

xaedes replied:
Ok, I changed the includes and used std::min & std::max instead of the macros.

xaedes committed:
c++ in baby-llama example

use c++ includes instead of c includes
use std::min, std::max instead of MIN, MAX macros

xaedes committed:
c++ in baby-llama example

use c++ includes instead of c includes
use std::min, std::max instead of MIN, MAX macros

ggerganov committed:
ggml : fix compiler warnings + cosmetic changes

ggerganov commented:
On M1 I get the following output:

<details>

```
$ ▶ make -j && ./bin/baby-llama 
Consolidate compiler generated dependencies of target ggml
[  3%] Generating build details from Git
-- Found Git: /opt/homebrew/bin/git (found version "2.39.0") 
[  6%] Building C object CMakeFiles/ggml.dir/ggml.c.o
[  6%] Built target BUILD_INFO
/Users/ggerganov/development/github/llama.cpp/ggml.c:9835:54: warning: implicit conversion increases floating-point precision: 'const float' to 'double' [-Wdouble-promotion]
                const float mean_eps = sum_xx/ne00 + eps;
                                                   ~ ^~~
/Users/ggerganov/development/github/llama.cpp/ggml.c:9836:52: warning: implicit conversion increases floating-point precision: 'float' to 'double' [-Wdouble-promotion]
                const float sum_eps  = sum_xx + eps*ne00;
                                              ~ ~~~^~~~~
/Users/ggerganov/development/github/llama.cpp/ggml.c:9942:55: warning: implicit conversion increases floating-point precision: 'const float' to 'double' [-Wdouble-promotion]
                ggml_vec_scale_f32(ne00, dx, -sum_xdz/sum_eps);
                                                     ~^~~~~~~
/Users/ggerganov/development/github/llama.cpp/ggml.c:9834:29: warning: unused variable 'mean' [-Wunused-variable]
                const float mean     = sum_xx/ne00;
                            ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:9837:29: warning: unused variable 'mean_xdz' [-Wunused-variable]
                const float mean_xdz = sum_xdz/ne00;
                            ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:9840:29: warning: unused variable 'rms' [-Wunused-variable]
                const float rms      = sqrtf(mean_eps);
                            ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:9842:29: warning: unused variable 'scale' [-Wunused-variable]
                const float scale    = -rrms/(ne00 * mean_eps); // -1/(n*rms**3)
                            ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:2771:20: warning: unused function 'ggml_vec_add1_f32' [-Wunused-function]
inline static void ggml_vec_add1_f32(const int n, float * z, const float * x, const float   v) { for (int i = 0; i < n; ++i) z[i]  = x[i] + v;    }
                   ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:2778:20: warning: unused function 'ggml_vec_mul_f32' [-Wunused-function]
inline static void ggml_vec_mul_f32 (const int n, float * z, const float * x, const float * y) { for (int i = 0; i < n; ++i) z[i]  = x[i]*y[i];   }
                   ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:3981:20: warning: unused function 'ggml_vec_silu_f16' [-Wunused-function]
inline static void ggml_vec_silu_f16(const int n, ggml_fp16_t * y, const ggml_fp16_t * x) {
                   ^
/Users/ggerganov/development/github/llama.cpp/ggml.c:4515:19: warning: unused function 'ggml_up64' [-Wunused-function]
static inline int ggml_up64(int n) {
                  ^
11 warnings generated.
[  6%] Built target ggml
[  9%] Linking CXX static library libllama.a
[ 12%] Built target llama
[ 21%] Linking CXX executable ../bin/test-quantize-perf
[ 21%] Linking CXX executable ../../bin/quantize
[ 24%] Linking CXX executable ../bin/test-quantize-fns
[ 24%] Linking CXX executable ../bin/test-tokenizer-0
[ 30%] Linking CXX executable ../bin/test-sampling
[ 30%] Linking CXX executable ../../bin/quantize-stats
[ 33%] Built target common
Consolidate compiler generated dependencies of target baby-llama
[ 36%] Linking CXX executable ../../bin/benchmark
[ 39%] Building CXX object examples/baby-llama/CMakeFiles/baby-llama.dir/baby-llama.cpp.o
[ 42%] Linking CXX executable ../../bin/embedding
[ 45%] Linking CXX executable ../../bin/vdot
[ 48%] Linking CXX executable ../../bin/perplexity
[ 51%] Linking CXX executable ../../bin/main
[ 54%] Linking CXX executable ../../bin/save-load-state
[ 57%] Linking CXX executable ../../bin/q8dot
[ 60%] Built target quantize-stats
[ 63%] Built target test-sampling
[ 66%] Built target test-tokenizer-0
[ 69%] Built target test-quantize-perf
[ 72%] Built target quantize
[ 75%] Built target test-quantize-fns
[ 78%] Built target benchmark
[ 81%] Built target embedding
[ 87%] Built target save-load-state
[ 87%] Built target perplexity
[ 90%] Built target main
[ 93%] Built target vdot
[ 96%] Built target q8dot
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:368:20: warning: unused variable 'n_embd' [-Wunused-variable]
    const uint32_t n_embd  = hparams.n_embd;
                   ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:370:20: warning: unused variable 'n_vocab' [-Wunused-variable]
    const uint32_t n_vocab = hparams.n_vocab;
                   ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:372:14: warning: unused variable 'n_ff' [-Wunused-variable]
    uint32_t n_ff = ((2*(4*hparams.n_embd)/3 + hparams.n_mult - 1)/hparams.n_mult)*hparams.n_mult;
             ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:401:20: warning: unused variable 'n_embd' [-Wunused-variable]
    const uint32_t n_embd  = hparams.n_embd;
                   ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:403:20: warning: unused variable 'n_vocab' [-Wunused-variable]
    const uint32_t n_vocab = hparams.n_vocab;
                   ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:405:14: warning: unused variable 'n_ff' [-Wunused-variable]
    uint32_t n_ff = ((2*(4*hparams.n_embd)/3 + hparams.n_mult - 1)/hparams.n_mult)*hparams.n_mult;
             ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:749:15: warning: unused variable 'n_lora' [-Wunused-variable]
    const int n_lora  = hparams.n_lora;
              ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:1190:30: warning: unused variable 'before_opt_best_samples' [-Wunused-variable]
        struct ggml_tensor * before_opt_best_samples = ggml_new_tensor_1d(ctx0, GGML_TYPE_I32, n_tokens);
                             ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:1191:30: warning: unused variable 'before_opt_probs' [-Wunused-variable]
        struct ggml_tensor * before_opt_probs        = ggml_new_tensor_2d(ctx0, GGML_TYPE_F32, n_vocab, n_tokens);
                             ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:1104:14: warning: unused parameter 'argc' [-Wunused-parameter]
int main(int argc, char ** argv) {
             ^
/Users/ggerganov/development/github/llama.cpp/examples/baby-llama/baby-llama.cpp:1104:28: warning: unused parameter 'argv' [-Wunused-parameter]
int main(int argc, char ** argv) {
                           ^
11 warnings generated.
[100%] Linking CXX executable ../../bin/baby-llama
[100%] Built target baby-llama
init model
init_kv_cache
Example 1
error_before_opt: 1390.99
error_after_opt:  758.28
best samples after optimization:
      X 
      X 
      X 
     X  
   X    
      X 
       X
  X     
Example 9
error_before_opt: 215.06
error_after_opt:  98.55
Example 17
error_before_opt: 90.67
error_after_opt:  82.54
Example 25
error_before_opt: 84.57
error_after_opt:  33.18
Example 33
error_before_opt: 45.32
error_after_opt:  27.95
Example 41
error_before_opt: 61.21
error_after_opt:  37.23
Example 49
error_before_opt: 48.33
error_after_opt:  28.40
Example 57
error_before_opt: 46.73
error_after_opt:  32.73
Example 65
error_before_opt: 52.76
error_after_opt:  33.15
best samples after optimization:
   X    
    X   
   X    
       X
       X
       X
       X
       X
Example 73
error_before_opt: 42.79
error_after_opt:  28.59
Example 81
error_before_opt: 38.62
error_after_opt:  26.49
Example 89
error_before_opt: 51.27
error_after_opt:  29.71
Example 97
error_before_opt: 41.93
error_after_opt:  27.56
Example 105
error_before_opt: 45.53
error_after_opt:  29.05
Example 113
error_before_opt: 38.62
error_after_opt:  22.80
Example 121
error_before_opt: 34.92
error_after_opt:  25.99
Example 129
error_before_opt: 35.48
error_after_opt:  24.03
best samples after optimization:
 X      
 X      
    X   
       X
      X 
       X
       X
       X
Example 137
error_before_opt: 43.82
error_after_opt:  29.17
Example 145
error_before_opt: 41.60
error_after_opt:  26.46
Example 153
error_before_opt: 37.05
error_after_opt:  26.87
Example 161
error_before_opt: 45.23
error_after_opt:  32.64
Example 169
error_before_opt: 47.27
error_after_opt:  36.41
Example 177
error_before_opt: 46.18
error_after_opt:  30.27
Example 185
error_before_opt: 49.92
error_after_opt:  30.48
Example 193
error_before_opt: 48.16
error_after_opt:  30.97
best samples after optimization:
 X      
 X      
 X      
   X    
    X   
     X  
       X
       X
Example 201
error_before_opt: 40.04
error_after_opt:  28.61
Example 209
error_before_opt: 40.37
error_after_opt:  26.17
Example 217
error_before_opt: 41.98
error_after_opt:  26.32
Example 225
error_before_opt: 50.48
error_after_opt:  29.77
Example 233
error_before_opt: 50.04
error_after_opt:  27.14
Example 241
error_before_opt: 49.34
error_after_opt:  26.62
Example 249
error_before_opt: 42.56
error_after_opt:  25.26
Generating 128 tokens.
X       
 X      
  X     
    X   
     X  
      X 
---
       X
       X
       X
      X 
     X  
   X    
       X
       X
       X
       X
 X      
 X      
      X 
     X  
      X 
     X  
 X      
  X     
 X      
 X      
 X      
 X      
  X     
   X    
    X   
     X  
      X 
       X
       X
       X
       X
      X 
     X  
   X    
       X
       X
       X
       X
 X      
 X      
      X 
     X  
      X 
     X  
 X      
  X     
 X      
 X      
 X      
 X      
  X     
   X    
    X   
     X  
      X 
       X
       X
       X
       X
      X 
     X  
   X    
       X
       X
       X
       X
 X      
 X      
      X 
     X  
      X 
     X  
 X      
  X     
 X      
 X      
 X      
 X      
  X     
   X    
    X   
     X  
      X 
       X
       X
       X
       X
      X 
     X  
   X    
       X
       X
       X
       X
 X      
 X      
      X 
     X  
      X 
     X  
 X      
  X     
 X      
 X      
 X      
 X      
  X     
   X    
    X   
     X  
      X 
       X
       X
       X
       X
      X 
     X  
   X    
       X
       X
       X
       X
 X      
 X      
      X 
     X  
      X 
     X  
 -0.86 0.50 -0.18 -0.09 -1.01 0.56 0.19 -0.47 -0.75 -1.35 0.56 0.69 -1.24 -0.68 0.85 -0.91 -0.16 0.04 0.06 -0.81 1.17 -1.11 -0.51 0.28 0.15 1.41 0.36 0.90 -1.05 -0.24 -0.95 -1.13
 1.02 1.37 0.90 -0.89 -1.07 -0.10 -0.05 0.88 0.81 -0.74 -0.33 -0.48 -0.26 0.15 -0.57 1.22 -0.01 0.61 0.67 1.00 1.00 0.19 -0.92 0.96 0.01 1.14 -0.40 0.82 1.24 1.00 -0.50 0.26
 -1.04 0.10 -0.84 1.04 0.35 1.46 -0.32 1.26 0.29 -0.62 1.09 0.62 -0.11 1.00 0.89 0.38 0.55 -0.48 0.32 1.04 -0.14 0.97 -1.27 0.33 0.46 0.11 -0.70 -0.63 -0.84 0.27 0.26 -0.94
 0.10 0.44 1.18 0.43 -1.13 0.16 1.21 -0.37 0.73 0.10 -1.02 0.45 -0.37 0.74 0.33 0.56 -1.53 -0.76 -0.10 0.43 -0.53 0.09 -1.32 -1.19 -0.48 0.91 -0.72 0.91 0.53 -0.51 0.09 -1.06
 -0.14 -0.05 1.31 0.49 0.56 -0.31 1.34 0.46 0.69 -0.06 -0.46 -0.96 -0.61 0.69 1.03 -0.31 0.12 -0.15 -0.82 -0.91 1.17 -0.33 0.09 -0.77 0.80 0.19 -1.78 -0.35 -0.76 -1.00 0.18 0.97
 -0.46 -0.32 0.93 1.02 0.46 -0.54 -0.90 0.65 0.81 -0.04 0.60 -0.38 -0.33 -0.57 -0.84 -1.00 -1.31 1.23 -0.78 0.78 0.71 -0.96 -0.18 -1.05 -0.41 0.14 -0.83 -0.06 -0.60 -1.45 -0.58 -0.30
 0.90 -0.11 0.67 -0.46 -0.86 -0.37 -0.93 -0.73 0.26 0.49 -0.56 -1.21 0.51 0.29 0.30 -0.82 -0.54 0.85 -0.13 -0.40 -0.27 -0.75 -0.41 -0.48 0.27 -0.48 -0.79 0.31 1.12 -0.97 0.91 -0.85
 0.31 0.38 -0.97 1.29 1.26 0.40 -0.75 0.81 0.44 -0.84 -0.83 -0.70 -0.50 0.70 1.17 0.82 0.88 1.36 0.03 -1.31 -0.80 0.41 0.67 0.22 -0.22 1.17 1.89 1.04 1.44 -0.82 -0.37 0.74
done
```
</details>

When I change the following `float`s to `ggml_float`s, the sinusoid completely disappears - i.e. it outputs straight line:

https://github.com/xaedes/llama.cpp/blob/78af3e92c9921f72956f7ae13a1a0f67134780af/ggml.c#L9834-L9842

Any idea why that might be?
Seems like there is some numerical instability there.

ggerganov committed:
ggml : fix nullptr derefs in GGML_OP_CONT and GGML_OP_RESHAPE back

xaedes commented:
> On M1 I get the following output

Hm, when I change them to ggml_float it still works well.
But some of the ggml_operations had `#ifdefs` for GGML_USE_ACCELERATE

For example in `ggml_compute_forward_add1_f32` 
https://github.com/ggerganov/llama.cpp/blob/6cc42deda511b5de82c43484b3e10be20fea224e/ggml.c#L8167-L8172

Maybe setting the parameter for src1 to 0 does not work the intended way? 

test_grad0.c also tests the forward pass a bit because of the finite differences method. Does that report anything unusual?
rms_norm sometimes gives slightly different gradient than finite differences when the gradient are big, but rel error is still small then.

xaedes committed:
swap arguments to vDSP_vdiv call

documentation for vDSP_vdiv states: "Note that B comes before A!"

xaedes committed:
swap arguments to vDSP_vdiv call

documentation for vDSP_vdiv states: "Note that B comes before A!"

ggerganov committed:
ggml : swap vDSP_vsub args as per documentation

ggerganov commented:
After fixing the `vDSP_vsub` argument order, it works now

![image](https://user-images.githubusercontent.com/1991296/236900501-6be7a1eb-f12b-4da8-ba11-cde45c4e9170.png)

xaedes committed:
add parallel batched forward function for baby-llama training

xaedes committed:
cleanup code for batched training

xaedes commented:
I got batched a forward function working. With this greater number of parallel batches can be trained with ease.

This should also be useful to implement beam search sampling.

xaedes committed:
remove trailing whitespace

ggerganov commented:
Ha, that is interesting. I somehow thought that to support batched inference some changes in `ggml.c` would be needed.
Well done!

> This should also be useful to implement beam search sampling.

Yes, this is needed for the beam-search decoding in `whisper.cpp`
Also, another project that can greatly benefit from batched inference is [bert.cpp](https://github.com/skeskinen/bert.cpp)

ggerganov commented on ggml.c:
```diff
@@ -11322,11 +13633,44 @@ static void ggml_compute_backward(struct ggml_context * ctx, struct ggml_tensor
+                    struct ggml_tensor* F08 = ggml_transpose (ctx, F07);
+                    struct ggml_tensor* F09 = ggml_cont      (ctx, F08);
+                    struct ggml_tensor* F10 = ggml_reshape   (ctx, F09, src0->grad);
+
```
Wow!

Btw, would it make sense to have `GGML_OP_REPEAT_BACK` that implements this in a kernel?
Maybe if this is some sort of bottleneck. Otherwise, keep it like this.

xaedes replied:
Yea I shuddered myself at this^^ I think one of the conts is not even necessary.
I also think it would make sense to implement this as an extra operation, should not be that difficult.
Most of the reshaping etc is only necessary to use sum_rows which could be done better in a special operation.

ggerganov committed:
minor : fix compiler warnings + indentation style

ggerganov committed:
ggml : fix null ptr deref in backward pass

ggerganov commented:
Hmm, what's wrong with the CI?
It was successful before my changes, then I made some minor edits (fix warnings, indentation) and now it is failing.

```
/home/runner/work/llama.cpp/llama.cpp/ggml.c: In function ‘ggml_compute_forward_add1’:
/home/runner/work/llama.cpp/llama.cpp/ggml.c:7398:14: error: ‘GGML_TYPE_Q4_2’ undeclared (first use in this function); did you mean ‘GGML_TYPE_Q4_1’?
 7398 |         case GGML_TYPE_Q4_2:
      |              ^~~~~~~~~~~~~~
      |              GGML_TYPE_Q4_1
```
Did it checkout the wrong branch 🤔

ggerganov commented:
@xaedes Is it ok if I merged latest `master` into your branch? In case you have some pending changes, you can do it so I don't mess up your flow

xaedes commented:
> Is it ok if I merged latest `master` into your branch? In case you have some pending changes, I can wait for you to do it yourself so I don't mess up your flow

@ggerganov Sure

ggerganov committed:
Merge remote-tracking branch 'origin/master' into HEAD

ggerganov committed:
ggml : remove Q4_2 remnants

ggerganov committed:
ggml : fix clang-tidy warnings

ggerganov committed:
baby-llama : couple of clang-tidy warnings

ggerganov reviewed (approved):
This is a pretty big addition to `ggml` - I hope it will open up some applications for training in the future

- The `tests` will be moved to the `ggml` repo after we merge this PR
- The batched forward example is very valuable. Have to apply it to `whisper.cpp` and `bert.cpp`
- After merging this, it would be a good time for some refactoring passes in `ggml.c` to try and reduce code duplication and overall code size

Thanks @xaedes - excellent effort!

RonanKMcGovern commented:
@xaedes has there been further progress in this direction since June? In particular, how far are things from being able to train LoRA adapters?

xaedes commented:
@RonanKMcGovern 
Yep, finetuning LORA adapters on LLAMA models works: https://github.com/ggerganov/llama.cpp/pull/2632
